File: solutions/0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        std::unordered_set<int>st;
        int left=0;
        int maxr=0;
        for(int right=0;right<s.size();right++){
            if(!st.contains(s[right]))
                st.insert(s[right]);
            else 
            {
                while(st.contains(s[right])){
                    st.erase(s[left]);
                    left++;
                    }st.insert(s[right]);
            }
            maxr=std::max(maxr,right-left+1);
        }
    return maxr;
    }
};

File: solutions/0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0;
        int left = 0;
        HashMap<Character, Integer> mp = new HashMap<>();

        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);

            if (mp.containsKey(c) && mp.get(c) >= left) {
                left = mp.get(c) + 1;
            }

            mp.put(c, right);
            max = Math.max(max, right - left + 1);
        }

        return max;
    }
}
